### PR TITLE
Add chance-aware detection metrics

### LINKED ISSUE #736: evaluate_print and the time-series tests do not catch a non-discriminating detector
`evaluate_print` reports ROC and precision@n with no chance level, so a number that looks reasonable can be sitting on the floor and nothing in the output says so.

At 5% contamination a random-scoring detector prints roughly:

```
dummy ROC:0.5244, precision @ rank n:0.12
```

0.12 reads as a low but real number. The floor is 0.05. A constant-scoring detector prints `precision @ rank n:0.0` and emits a silent sklearn `UndefinedMetricWarning`.

The test does not catch it because it asserts nothing:

https://github.com/yzhao062/pyod/blob/master/pyod/test/test_data.py#L470

It passes `y_train * 0.1`, which is the labels scaled, so the input is a perfect oracle by construction.

The seven time-series test files have the same shape. Each binds `self.y_train, self.y_test` in `setUp` and then never references them again:

https://github.com/yzhao062/pyod/blob/master/pyod/test/test_ts_spectral_residual.py#L20

The assertions are on length, non-negativity and NaN. A detector returning `np.ones(n)` passes all of them, at lift 1.00 against the fixture's own prevalence. I checked five of the seven that way.

You already wrote the guard for this, in one place:

https://github.com/yzhao062/pyod/blob/master/pyod/test/test_deepsvdd.py#L65

The comment there is exactly the point, that "a ROC floor can still be met by floating-point noise in the ordering". It runs for one detector out of ninety-four test files.

Proposed, and I am happy to open the PR:

1. `detection_lift()` in `pyod/utils/utility.py`, exported. Observed precision@n over the contamination floor.
2. `evaluate_print` prints the chance level alongside precision@n, so the floor travels with the number.
3. Real assertions in `test_evaluate_print`.
4. One shared helper the seven time-series tests call, asserting the scores separate the labels those fixtures already carry.

Under 200 lines, no API break, defaults unchanged. It goes in `utils` rather than `test` because `pyod/test/` is excluded from the wheel, so users of the library get the floor too.

Background on why the floor moves with prevalence and why a threshold-independent number alone does not settle it: https://doi.org/10.1109/ACCESS.2026.3705430

### PR DESCRIPTION
**Pull request by be-student**
Closes #736.

## Problem

`evaluate_print` reports precision at rank n without its chance-level baseline, so the same precision value can mean very different things at different outlier prevalences. The existing evaluation test also passes scaled ground-truth labels as scores, while the seven time-series detector tests did not verify that their scores discriminate the fixture labels.

## Changes

- Add and publicly export `detection_lift`, defined as precision at rank n divided by outlier prevalence.
- Return `0.0` when the ground truth has no outliers and reject empty ground truth with `ValueError`.
- Print chance level and detection lift from `evaluate_print` alongside ROC and precision at rank n.
- Make `precision_n_scores` return zero without an `UndefinedMetricWarning` when no positives are predicted.
- Add real assertions for `evaluate_print` output and metric edge cases.
- Add one shared quality assertion to all seven time-series fixtures. It rejects scores that become constant after rounding to eight decimal places, then requires `detection_lift > 1.1`.
- Align the KShape, MatrixProfile, and SAND quality fixtures with their subsequence semantics: they now contain mixed point/subsequence anomalies and use a 10-sample scoring window, matching the generator's 10-sample subsequences.
- Normalize `pyod/utils/data.py` line endings to avoid carrying mixed CRLF/LF content forward.

## CI failure and fix

The previous remote head, `f3c4e863434bfef43747528da68eedaf625541dd`, failed the same three tests on every Linux, Windows, and macOS matrix job:

- `TestKShape::test_fit`
- `TestMatrixProfile::test_fit`
- `TestSAND::test_fit`

Those detectors score subsequences and map each window score back across its timestamps. Their former point-only fixtures used 20-sample windows, which created broad tied score plateaus. Precision at the true anomaly count could therefore be zero even though the scores were nonconstant. The metric and reviewer-requested guard were kept unchanged; the fixtures were corrected to exercise detectable subsequences at the detector's native resolution.

At final local head `4696fbe0b9fb1b02d2e6492f9c5c58acf694d412`, the exact-label lifts are:

| Detector | Detection lift | Rounded unique scores |
| --- | ---: | ---: |
| KShape | 6.750000 | 92 |
| MatrixProfile | 10.909091 | 78 |
| SAND | 10.227273 | 133 |

All exceed the shared `1.1` margin and remain nonconstant after rounding to eight decimal places.

## Validation

- macOS, Python 3.9.6: affected test files, `19 passed`
- macOS, Python 3.12.13: affected test files, `19 passed`
- macOS, Python 3.13.7: affected test files, `19 passed`
- Python 3.13.7 PR-focused suite: `70 passed, 7 skipped, 8 subtests passed`
- `git diff --check`
- Clean synthetic merge with current upstream `master` at `3a71f47bb2cb9b7c8a09fc92e5c67510d7198b58`
- Hosted CI at `4696fbe` passed all eleven Linux, Windows, and macOS matrix jobs, the packaging smoke test, and CodeQL.

The seven local skips are the optional PyTorch-backed tests because PyTorch was not installed in the local validation environment. Hosted CI executed the full dependency set successfully on the corrected head.

Python 3.9 emits existing KShape matrix-multiplication runtime warnings with both the original and corrected fixtures; they are unrelated to this change and do not fail the tests.

## AI assistance disclosure

OpenAI Codex assisted with implementation, CI-log analysis, review-feedback remediation, and test execution. The resulting diff and validation evidence were checked before publication; no manual testing is claimed.

### CODE REVIEW

**Review by mateenali66 — COMMENTED** [review]
the shared helper is weaker than the deepsvdd guard it generalises, and it lets the degenerate case through about half the time.

`test_deepsvdd.py::test_scores_are_not_constant` counts uniques after `np.round(scores, 8)`, and the comment there says why, "a ROC floor can still be met by floating-point noise in the ordering". `assert_scores_separate_labels` drops the rounding and then uses `auc > 0.5`, which is that floor.

constant detector, 300 samples, 10% contamination:

```python
s = np.full(n, 0.5) + rng.normal(0, 1e-17, n)
```

`np.unique(s).size` is 2, since float64 spacing near 0.5 is about 1.1e-16 and the noise nudges a couple of samples onto the next representable value. that clears `> 1`. auc then falls either side of 0.5 by luck. across 2000 seeds `assert_scores_separate_labels` accepts this detector on 970 of them, 48.5%. the deepsvdd guard rejects all 2000, because rounding to 8dp collapses it back to one unique value.

the fix is already in this PR. assert `detection_lift` with a margin instead of `auc > 0.5`. lift is prevalence-baselined, so one threshold means the same thing across all seven fixtures, and it is the metric the PR is adding. carrying over the `np.round(scores, 8)` from deepsvdd closes the constant check.

File: pyod/utils/data.py [hunk 1]
```diff
@@ -19,8 +19,9 @@
 from sklearn.utils import check_random_state
 from sklearn.utils import column_or_1d
 
-from .utility import check_parameter
-from .utility import precision_n_scores
+from .utility import check_parameter
+from .utility import detection_lift
```
Comment (cindysteward) [inline pyod/utils/data.py:23]: noticed this file isCRLF throughout, so every untouched or removed line carries a trailing \r. Butthe 3 new import lines here and the new lines further down in evaluate_print are LF-only. Worth normalising, otherwise the diff carries mixed line endings forward, future edits to this file could show spurious changes on every line

**Comment by cindysteward** [comment]
> the shared helper is weaker than the deepsvdd guard it generalises, and it lets the degenerate case through about half the time.
> 
> `test_deepsvdd.py::test_scores_are_not_constant` counts uniques after `np.round(scores, 8)`, and the comment there says why, "a ROC floor can still be met by floating-point noise in the ordering". `assert_scores_separate_labels` drops the rounding and then uses `auc > 0.5`, which is that floor.
> 
> constant detector, 300 samples, 10% contamination:
> 
> ```python
> s = np.full(n, 0.5) + rng.normal(0, 1e-17, n)
> ```
> 
> `np.unique(s).size` is 2, since float64 spacing near 0.5 is about 1.1e-16 and the noise nudges a couple of samples onto the next representable value. that clears `> 1`. auc then falls either side of 0.5 by luck. across 2000 seeds `assert_scores_separate_labels` accepts this detector on 970 of them, 48.5%. the deepsvdd guard rejects all 2000, because rounding to 8dp collapses it back to one unique value.
> 
> the fix is already in this PR. assert `detection_lift` with a margin instead of `auc > 0.5`. lift is prevalence-baselined, so one threshold means the same thing across all seven fixtures, and it is the metric the PR is adding. carrying over the `np.round(scores, 8)` from deepsvdd closes the constant check.

reproduced this with afresh RNG (numpy Generator, 2000 seeds), same setup, n=300, 10% contamination, noise at 1e-17 around 0.5. Raw np.unique(scores).size > 1 passed 1114/2000, 55.7%. Rounding to 8dp first passed 0/2000, matching the deepsvdd guard. Numbers differ a bit from yours, likely RNG choice, mechanism and conclusion both hold. So yes, assert detection_lift with a margin plus the round(scores, 8) step, it's prevalence-baselined so one threshold means the same thing across all 7

**Review by mateenali66 — COMMENTED** [review]
The lift bound lets a random scorer through. Pure random scores clear > 1.1 on 54.4% of runs against the four n=300 fixtures and 36.0% against the three n=500 ones, 20k trials each. Lift is quantised by the outlier count, so > 1.1 is really "one hit in the top 15", or two in the top 25.

Measured what the detectors actually score: ts_od 3.0, kshape 6.8, sand 10.2, matrix_profile 10.9, and spectral_residual, lstm and anomaly_transformer all 20.0. So > 2.7 passes all seven and drops random to about 3%.

ts_od at 3.0 is the floor and 2.7 sits close under it. Raise the bound, or grow the fixtures so a bound has somewhere to sit?

The round(scores, 8) half is doing its job. Constant detector 0 of 2000, against 1109 and 1471 before.
